### PR TITLE
Switch done philosophy for trajectory forwarding

### LINKED ISSUE #17: forward controllers should wait for a signal from the hardware to finish
Currently, the passthrough controllers send the trajectory and when the designated time is up, it is checked whether the robot is at its final position or not (within specified limits).

I think that we should not use such a fire&forget&check interface but rather wait for the underlying hardware to signal completion.

E.g. for the `ur_robot_driver` we should send something back when reaching [this line](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/blob/beta-testing/ur_robot_driver/resources/ros_control.urscript#L153).

As this will introduce a potential endless wait, we should implement a strategy to either read keepalive signals from the hardware or add a user-specified timeout that we can use ontop of the trajectory time to declare a goal `LOST´.

### PR DESCRIPTION
**Pull request by stefanscherzinger**
**Problem**
As pointed out by @fmauch [here](#17) and with the help of [these plots](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/pull/1#issuecomment-549894865), the robot drivers' feedback (for the Universal Robots in this case) might not be concise enough for open-loop trajectory passthrough when being in _protective stop_ or similar blocking states. We need to inform the `PassThroughController`'s action client that the trajectory is not being executed as expected.

**Approach**
- The RobotHW now declares the trajectory execution done by calling a special method in the trajectory interface.
- The passthrough controllers themselves only warn when the time is up but do not stop trajectory execution.

Fixes #17

### CODE REVIEW

**Comment by stefanscherzinger** [comment]
@fmauch What do you think? I'm not too happy to introduce [yet another error Flag](https://github.com/fzi-forschungszentrum-informatik/cartesian_ros_control/blob/add-done-callback-for-trajectory-passthrough/pass_through_controllers/include/pass_through_controllers/trajectory_interface.h#L58). We already have [JointTrajectoryResult](http://docs.ros.org/en/api/control_msgs/html/action/FollowJointTrajectory.html) and [CartesianTrajectoryResult](https://github.com/fzi-forschungszentrum-informatik/cartesian_ros_control/blob/master/cartesian_control_msgs/action/FollowCartesianTrajectory.action) that we must use to inform the action client. Unfortunately, they don't provide semantic fields for _passthrough failed due to something with the robot..._. The many mappings from driver library -> ROS -> Passthrough -> *ActionResult seem a little suboptimal.

Any suggestions?

---

Edit 1: We would call the trajectory interface's [setDone(...)](https://github.com/fzi-forschungszentrum-informatik/cartesian_ros_control/blob/add-done-callback-for-trajectory-passthrough/pass_through_controllers/include/pass_through_controllers/trajectory_interface.h#L240) method [here](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/pull/396/files#diff-ef278955f7d85bf066e8cbf3d9ba847432895b0025a10cb89c6d6c6dd0596aa9R1191) in the driver.

Edit 2: Looking at the `urcl::control::TrajectoryResult` you plan to use, I'm wondering if a `true/false` distinction could be sufficient. The flag for the user-initiated preemption (by trajectory client) is not necessary in my opinion, since the PassThroughController is already aware of that (it forwarded this request to the driver). Or are you referring to cancel via Teach-Pendant?

Edit 3: Thinking about it, we could also wait in the `PassThroughController` for the preemption request to actually succeed in the hardware before returning to the action client. I'm a little worried about all these possible _endless_ waits, though.

File: pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp [hunk 1]
```diff
@@ -158,10 +162,16 @@ namespace trajectory_controllers {
...
       monitorExecution(f);
 
-      // Time is up. Check goal tolerances and set terminal state.
+      // Time is up.
       if (action_duration_.current >= action_duration_.target)
       {
-        timesUp();
+        if (!done_)
+        {
+          ROS_WARN_ONCE(
```
Comment (fmauch) [inline pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp]: To my knowledge `ROS_WARN_ONCE` prints once per node lifetime. I guess, that is not what you want.

Comment (stefanscherzinger) [reply pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp]: It's now a throttled warning. See [here](https://github.com/fzi-forschungszentrum-informatik/cartesian_ros_control/pull/27/files#diff-94485e2c60bb3683c7c4dd8222ebee10d77fe263ac8fd976cc0219a9876e6970R155).

**Comment by stefanscherzinger** [comment]
After further discussions with @fmauch , this is the final result:

We introduce a design change for how the PassThroughControllers interact with the robot. They reject invalid trajectories like before. But when canceling running trajectory execution, e.g. by user preemption request, failing path tolerances, or unexpected errors on the robot side, they ultimately wait for the robot to confirm this with using the done callback mechanism.

This makes sure that the PassThroughControllers do not signal success or preemption to action clients while the robot is still moving or behaving unresponsive.

The TrajectoryInterface drops the previous handle mechanism in favor of a cleaner callback-only approach.

**Review by fmauch — APPROVED** [review]
Tested with integration into ur_robot_driver.

**Pull request merged**